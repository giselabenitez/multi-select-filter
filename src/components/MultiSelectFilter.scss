@use "sass:color";

$primary-color: #3063d0;
$border-color: #ddd;
$bg-color: #f8f8f8;

.filter-container {
  background: $bg-color;
  border: 1px solid $border-color;
  border-radius: 6px;
  padding: 16px;
  width: 330px;
  font-family: Arial, sans-serif;

  h3 {
    font-size: 16px;
    font-weight: lighter;
    margin-bottom: 12px;
    text-align: left;
  }
}

.search-bar {
  display: flex;
  align-items: center;
  border: 1px solid $border-color;
  border-radius: 4px;
  padding: 6px;
  background: white;

  input {
    font-size: 15px;
    font-weight: lighter;
    flex: 1;
    border: none;
    outline: none;
    padding: 6px;
  }

  img {
    width: 16px;
    height: 16px;
    margin-left: 6px;
  }
}

.filtered-list {
  height: 200px;
  width: 325px;
  overflow-y: auto;
  background: $bg-color;
  margin-top: 15px;
  margin-bottom: 15px;

  .list-item {
    display: flex;
    align-items: center;
    gap: 6px;
    margin-bottom: 6px;

    input[type="checkbox"] {
      accent-color: $primary-color;
    }
  }
}

.apply-button {
  width: 100%;
  padding: 10px;
  margin-top: 10px;
  color: white;
  background-color: $primary-color;
  font-size: 18px;
  font-weight: lighter;
  border: none;
  border-radius: 5px;
  cursor: pointer;

  &:hover {
    background-color: color.adjust($primary-color, $lightness: -10%);
  }
}