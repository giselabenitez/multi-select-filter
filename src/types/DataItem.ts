export interface DataItem {
    name: string;
    checked: boolean;
}