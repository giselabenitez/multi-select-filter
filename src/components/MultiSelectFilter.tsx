import {useEffect, useState} from "react";
import SearchBar from "./SearchBar.tsx";
import FilteredList from "./FilteredList.tsx";
import {DataItem} from "../types/DataItem.ts";
import "./MultiSelectFilter.scss";

const decodeHTML = (text: string) => {
    return new DOMParser().parseFromString(text, "text/html").body.textContent || "";
};

const MultiSelectFilter = () => {
    const [data, setData] = useState<DataItem[]>([]);
    const [filterText, setFilterText] = useState("");
    const [filtered, setFiltered] = useState<DataItem[]>([]);
    const [selected, setSelected] = useState<DataItem[]>([]);

    const handleChecked = (item: DataItem) => {
        const updatedData: DataItem[] = data.map((i) => item.name === i.name ? {...i, checked: !i.checked} : i);
        setData(updatedData);
    }

    useEffect(() => {
        fetch("/items.json")
            .then((items) => items.json())
            .then((items) => items.data.map((item: string) => ({name: decodeHTML(item), checked: false} as DataItem)))
            .then((data) => setData(data))
            .catch((error) => console.error("Error fetching JSON:", error));
    }, []);

    useEffect(() => {
        const filteredData = data
            .filter((item) => item.name.toLowerCase().includes(filterText.toLowerCase()))
            .filter((item) => !item.checked);
        setFiltered(filteredData);
        setSelected(data.filter(item => item.checked));
    }, [data, filterText]);

    return (
        <div className={"filter-container"}>
            <h3>Productgroep</h3>
            <SearchBar filterText={filterText} onTextChange={setFilterText}/>
            <FilteredList data={filtered} checkedList={selected} onChecked={handleChecked}/>
            <button className="apply-button">Toepassen</button>
        </div>
    );
};

export default MultiSelectFilter;