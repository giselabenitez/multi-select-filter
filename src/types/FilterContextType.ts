import {DataItem} from "./DataItem";

export interface FilterContextType {
    data: DataItem[];
    filterText: string;
    setFilterText: (text: string) => void;
    handleChecked: (item: DataItem) => void;
}