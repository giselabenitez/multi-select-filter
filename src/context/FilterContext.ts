import {createContext} from "react";
import {FilterContextType} from "../types/FilterContextType.ts";

export const FilterContext = createContext<FilterContextType | undefined>(undefined);
