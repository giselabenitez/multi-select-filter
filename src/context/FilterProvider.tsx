import * as React from "react";
import {useEffect, useState} from "react";
import {DataItem} from "../types/DataItem.ts";
import {FilterContext} from "./FilterContext.ts";

const decodeHTML = (text: string) => {
    return new DOMParser().parseFromString(text, "text/html").body.textContent || "";
};

export const FilterProvider: React.FC<{ children: React.ReactNode }> = ({children}) => {
    const [data, setData] = useState<DataItem[]>([]);
    const [filterText, setFilterText] = useState("");

    useEffect(() => {
        const storedData = localStorage.getItem("dataItems");
        if (storedData) {
            setData(JSON.parse(storedData));
        } else {
            fetch("/items.json")
                .then((items) => items.json())
                .then(({data}) => setData(data.map((item: string) => ({
                    name: decodeHTML(item),
                    checked: false
                } as DataItem))))
                .catch((error) => console.error("Error fetching JSON:", error));
        }
    }, []);

    const handleChecked = (item: DataItem) => {
        setData(prevState => prevState.map((i) =>
            item.name === i.name ? {...i, checked: !i.checked} : i
        ));
    };

    return (
        <FilterContext.Provider value={{data, filterText, setFilterText, handleChecked}}>
            {children}
        </FilterContext.Provider>
    );
}