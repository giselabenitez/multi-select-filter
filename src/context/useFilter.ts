import {FilterContextType} from "../types/FilterContextType.ts";
import {useContext} from "react";
import {FilterContext} from "./FilterContext.ts";

export const useFilter = (): FilterContextType => {
    const context = useContext(FilterContext);
    if (!context) {
        throw new Error('useFilter must be used within a FilterProvider');
    }
    return context;
};