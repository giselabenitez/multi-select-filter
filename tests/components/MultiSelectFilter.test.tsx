import React, {JSX} from "react";
import {beforeEach, describe, expect, test, vi} from "vitest";
import MultiSelectFilter from "../../src/components/MultiSelectFilter";
import {useFilter} from "../../src/context/useFilter";
import {FilterContextType} from "../../src/types/FilterContextType";
import {fireEvent, render, screen} from "@testing-library/react";

vi.mock("../../src/context/useFilter", () => ({
    useFilter: vi.fn<() => FilterContextType>(),
}));

describe("MultiSelectFilter", () => {
    let component: JSX.Element;

    beforeEach(() => {
        global.localStorage = {
            setItem: vi.fn(),
            getItem: vi.fn(),
            clear: vi.fn(),
            removeItem: vi.fn(),
            key: vi.fn(),
            length: 0,
        };

        component = <MultiSelectFilter/>;
    });

    test("renders without crashing when no data", async () => {
        useFilter.mockReturnValue({
            data: [],
            filterText: "",
            setFilterText: vi.fn(),
            handleChecked: vi.fn(),
        });

        render(component);

        expect(screen.getByText("Productgroep")).toBeInTheDocument();
    });

    test("renders with data and no filterText set", async () => {
        useFilter.mockReturnValue({
            data: [{name: "Thrillers", checked: false}, {name: "Fantasy", checked: false}],
            filterText: "",
            setFilterText: vi.fn(),
            handleChecked: vi.fn(),
        });

        const {asFragment} = render(component);

        expect(asFragment()).toMatchSnapshot();
        expect(screen.getByText("Thrillers")).toBeInTheDocument();
        expect(screen.getByText("Fantasy")).toBeInTheDocument();
    });

    test("renders with filter text", async () => {
        useFilter.mockReturnValue({
            data: [{name: "Kinderboeken", checked: false}, {name: "Kookboeken", checked: false}],
            filterText: "Kookboeken",
            setFilterText: vi.fn(),
            handleChecked: vi.fn(),
        });

        const {asFragment} = render(component);

        expect(asFragment()).toMatchSnapshot();
        expect(screen.queryByText("Kinderboeken")).not.toBeInTheDocument();
        expect(screen.queryByText("Kookboeken")).toBeInTheDocument();
    });

    test("filtered list is updated based on filter text", async () => {
        useFilter.mockReturnValue({
            data: [
                {name: "Psychologie", checked: false},
                {name: "Alle boeken", checked: false},
                {name: "Ebooks", checked: false},
                {name: "Alle muziek", checked: false}
            ],
            filterText: "All",
            setFilterText: vi.fn(),
            handleChecked: vi.fn(),
        });

        render(component);
        expect(screen.queryByText("Alle boeken")).toBeInTheDocument();
        expect(screen.queryByText("Alle muziek")).toBeInTheDocument();
        expect(screen.queryByText("Psychologie")).not.toBeInTheDocument();
        expect(screen.queryByText("Ebooks")).not.toBeInTheDocument();
        expect(screen.getAllByRole("checkbox")).toHaveLength(2);
    });

    test("handleChecked is called when the item is clicked", async () => {
        const mockHandleChecked = vi.fn();
        useFilter.mockReturnValue({
            data: [{name: "Film", checked: false}, {name: "Games", checked: false}],
            filterText: "",
            setFilterText: vi.fn(),
            handleChecked: mockHandleChecked,
        });

        const {asFragment} = render(component);

        expect(asFragment()).toMatchSnapshot();
        fireEvent.click(screen.getByText("Games"));
        expect(mockHandleChecked).toHaveBeenCalled();
        expect(mockHandleChecked).toHaveBeenCalledWith({name: "Games", checked: false});
    });

    test("updated data is saved in localStorage when 'Toepassen' button is clicked", async () => {
        const mockLocalStorage = {setItem: vi.fn()};
        global.localStorage = mockLocalStorage;

        useFilter.mockReturnValue({
            data: [{name: "Film", checked: true}],
            filterText: "",
            setFilterText: vi.fn(),
            handleChecked: vi.fn(),
        });

        render(<MultiSelectFilter/>);
        fireEvent.click(screen.getByRole("button", {name: "Toepassen"}));

        expect(mockLocalStorage.setItem).toHaveBeenCalledWith(
            "dataItems",
            JSON.stringify([{name: "Film", checked: true}])
        );
    });
});