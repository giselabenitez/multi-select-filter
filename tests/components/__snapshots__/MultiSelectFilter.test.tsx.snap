// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MultiSelectFilter > handleChecked is called when the item is clicked 1`] = `
<DocumentFragment>
  <div
    class="filter-container"
  >
    <h3>
      Productgroep
    </h3>
    <div
      class="search-bar"
    >
      <input
        placeholder="Zoek op..."
        type="text"
        value=""
      />
      <img
        alt="search"
        src="/search.svg"
      />
    </div>
    <div
      class="filtered-list"
    >
      <label
        class="list-item"
      >
        <input
          type="checkbox"
        />
        <span
          class="checkmark"
        />
        <span
          class="item-name"
        >
          Film
        </span>
      </label>
      <label
        class="list-item"
      >
        <input
          type="checkbox"
        />
        <span
          class="checkmark"
        />
        <span
          class="item-name"
        >
          Games
        </span>
      </label>
    </div>
    <button
      class="apply-button"
    >
      Toepassen
    </button>
  </div>
</DocumentFragment>
`;

exports[`MultiSelectFilter > renders with data and no filterText set 1`] = `
<DocumentFragment>
  <div
    class="filter-container"
  >
    <h3>
      Productgroep
    </h3>
    <div
      class="search-bar"
    >
      <input
        placeholder="Zoek op..."
        type="text"
        value=""
      />
      <img
        alt="search"
        src="/search.svg"
      />
    </div>
    <div
      class="filtered-list"
    >
      <label
        class="list-item"
      >
        <input
          type="checkbox"
        />
        <span
          class="checkmark"
        />
        <span
          class="item-name"
        >
          Thrillers
        </span>
      </label>
      <label
        class="list-item"
      >
        <input
          type="checkbox"
        />
        <span
          class="checkmark"
        />
        <span
          class="item-name"
        >
          Fantasy
        </span>
      </label>
    </div>
    <button
      class="apply-button"
    >
      Toepassen
    </button>
  </div>
</DocumentFragment>
`;

exports[`MultiSelectFilter > renders with filter text 1`] = `
<DocumentFragment>
  <div
    class="filter-container"
  >
    <h3>
      Productgroep
    </h3>
    <div
      class="search-bar"
    >
      <input
        placeholder="Zoek op..."
        type="text"
        value="Kookboeken"
      />
      <img
        alt="search"
        src="/search.svg"
      />
    </div>
    <div
      class="filtered-list"
    >
      <label
        class="list-item"
      >
        <input
          type="checkbox"
        />
        <span
          class="checkmark"
        />
        <span
          class="item-name"
        >
          Kookboeken
        </span>
      </label>
    </div>
    <button
      class="apply-button"
    >
      Toepassen
    </button>
  </div>
</DocumentFragment>
`;
